### DISCUSSION TITLE
HNSW clarification questions

### DISCUSSION THREAD

**Discussion by Madd0g** [cozodb/cozo#124] [Q&A] [answered]
Hi, trying out the new vector features, got some use cases to work somewhat, trying to understand what's happening

1. Scale of the different distance metrics - I see L2 and Cosine returning numbers in a different scale (L2 often returning numbers bigger than 1), is there a recommended way of setting up `radius` for the different metrics (considering the different scale)? I'm using sentence vectors
2. in the docs, when creating an index there's lot about connections and neighbors - if the index is for one table at a time, what are all these connections and nearest neighbors?
3. I copied the examples from the docs and got some of these errors:

thread '<unnamed>' panicked at 'index out of bounds: the len is 3 but the index is 3', cozo-core/src/runtime/hnsw.rs:960:25

I commented out these lines and the errors stopped:

```
#bind_vector: bv,
#bind_field: f,
#bind_field_idx: fi
```

Thought I'd mention it just because I've never seen this panic error before.

Thanks!

**Comment by zh217** [comment, marked as answer]
1. In general you should use Cosine, unless your vectors are already normalized, in which case Cosine and L2 should give almost identical results (but L2 will be faster). You can try to compute distance between vectors that you _want_ to be considered together in order to set the radius. There is no generally applicable rule about it.
2. The connections and neighbours refer to the internal graphs of the HNSW indices.
3. It is definitely a bug. The program should not panic no matter what you throw at it. I will look into it when I find time.
